api/job.go

```diff
@@ -0,0 +1,427 @@
+		}
+		return nil, err
+	}
+	return j, nil
+}
+
+func jobFilterByContext(contexts []permTypes.PermissionContext, filter *job.Filter) *job.Filter {
+	if filter == nil {
+		filter = &job.Filter{}
+	}
+contextsLoop:
+	for _, c := range contexts {
+		switch c.CtxType {
+		case permTypes.CtxGlobal:
+			filter.Extra = nil
+			break contextsLoop
+		case permTypes.CtxTeam:
+			filter.ExtraIn("teams", c.Value)
+		case permTypes.CtxJob:
+			filter.ExtraIn("name", c.Value)
+		case permTypes.CtxPool:
+			filter.ExtraIn("pool", c.Value)
+		}
+	}
+	return filter
+}
+
+// title: job list
+// path: /jobs/list
+// method: GET
+// produce: application/json
+// responses:
+//	200: List jobs
+//	204: No content
+//	401: Unauthorized
+func jobList(w http.ResponseWriter, r *http.Request, t auth.Token) error {
+	ctx := r.Context()
+	filter := &job.Filter{}
+	if name := r.URL.Query().Get("name"); name != "" {
+		filter.Name = name
+	}
+	if teamOwner := r.URL.Query().Get("teamOwner"); teamOwner != "" {
+		filter.TeamOwner = teamOwner
+	}
+	if owner := r.URL.Query().Get("owner"); owner != "" {
+		filter.UserOwner = owner
+	}
+	if pool := r.URL.Query().Get("pool"); pool != "" {
+		filter.Pool = pool
+	}
+	jobs, err := job.List(ctx, jobFilterByContext(permission.ContextsForPermission(t, permission.PermJobRead), filter))
+	if err != nil {
+		return err
+	}
+	if len(jobs) == 0 {
+		w.WriteHeader(http.StatusNoContent)
+		return nil
+	}
+	return json.NewEncoder(w).Encode(jobs)
+}
+
+// title: job trigger
+// path: /job/trigger/{name}
+// method: PUT
+// produce: application/x-json-stream
+// responses:
+//	200: OK
+//	401: Unauthorized
+//	404: Not found
+func jobTrigger(w http.ResponseWriter, r *http.Request, t auth.Token) (err error) {
+	ctx := r.Context()
+	name := r.URL.Query().Get(":name")
+	j, err := getJob(ctx, name)
+	if err != nil {
+		return err
+	}
+	canRun := permission.Check(t, permission.PermJobRun,
+		contextsForJob(j)...,
+	)
+	if !canRun {
+		return permission.ErrUnauthorized
+	}
+	evt, err := event.New(&event.Opts{
+		Target:     jobTarget(j.Name),
+		Kind:       permission.PermJobRun,
+		Owner:      t,
+		CustomData: event.FormToCustomData(InputFields(r)),
+		Allowed:    event.Allowed(permission.PermJobReadEvents, contextsForJob(j)...),
+	})
+	if err != nil {
+		return err
+	}
+	defer func() { evt.Done(err) }()
+	err = job.Trigger(ctx, j)
+	if err != nil {
+		return err
+	}
+	msg := map[string]interface{}{
+		"status": "success",
+	}
+	jsonMsg, err := json.Marshal(msg)
+	if err != nil {
+		return err
+	}
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	w.Write(jsonMsg)
+	return nil
+}
+
+// title: job info
+// path: /jobs
+// method: GET
+// produce: application/x-json-stream
+// responses:
+//	200: OK
+//	401: Unauthorized
+//	404: Not found
+func jobInfo(w http.ResponseWriter, r *http.Request, t auth.Token) (err error) {
```

Comment:
Why this route is implemented as a stream ?

Reply:
mistyped 😜 nice catch